### PR TITLE
Add debugger displays / type proxies for Timer

### PR DESCRIPTION
**Pull request by stephentoub**
Adds a private Timer.AllTimers property that can be used in the debugger to get a list of all timers, and debugger display strings for Timer / TimerQueueTimer that help to make sense of the state in a timer and when it'll fire next.

There's one oddity here that I worked around but it leads to another (IMHO less bad oddity).  The VS debugger does lazy evaluation of things like debugger displays, so if you have a long list of something with displays, it'll only evaluate them when they're in view.  The most relevant piece of information about a timer is when it will next fire, and computing that from the state in the timer requires getting the current tick count.  Thus, the debugger display for a timer needs to fetch TickCount, which obviously depends on the moment its read.  If you then had a list of 1000 timers but only 10 were in view, only those 10 would have their displays calculated, based on the current tick count.  Time marches on and you scroll down, exposing another 10, at which point their displays are calculated based on the now current TickCount.  This leads to weird inconsistencies in the view.  To work around that, I have the Timer.AllTimers property configured not to return a Timer or TimerQueueTimer, but actually an instance of the debugger type proxy, which I've also enabled to store a "current" tick count; that way, I can eagerly create all of the proxies and imbue each with the same view of time.  This mostly solves the aforementioned problem, but at the expense of the type column saying the instance is a TimerDebuggerTypeProxy rather than a Timer or TimerQueueTimer.  For now given the tradeoff, I think that's ok.  There are other inconsistencies, e.g. if you navigate from the proxy to the actual wrapped timer and view its display, since it's then factoring in a different "now", it'll show different results.  Not much we can do about it.  And this is all expert-level hidden away anyway, since Timer.AllTimers is private.

image

cc: @davidfowl, @r-ramesh, @kouvel

### CODE REVIEW

**Comment by r-ramesh** [comment]
@stephentoub This will cause a lot of inconsistencies in the debugger. The same Timer viewed in different contexts will yield different results (even adding one as a Watch). Is it not possible to merely show the dueTime as an absolute rather than relative time?

**Comment by stephentoub** [comment]
> This will cause a lot of inconsistencies in the debugger

Right,  that's what I was explaining. 

> Is it not possible to merely show the dueTime as an absolute rather than relative time?

We could.  I think that's harder to reason about for a developer,  having to do math in one's head to know how long until a timer fires. There will also be some of the same inconsistencies. But we could.

**Comment by stephentoub** [comment]
I changed it to use DateTime for the next due time.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs [hunk 1]
```diff
@@ -444,6 +493,11 @@ internal TimerQueueTimer(TimerCallback timerCallback, object? state, uint dueTim
                 Change(dueTime, period);
         }
 
+        internal string DisplayString =>
+            "Due = " + (_dueTime == Timeout.UnsignedInfinite ? "(not set)" : (DateTime.Now + TimeSpan.FromMilliseconds(_dueTime - (TimerQueue.TickCount64 - _startTicks)))) + ", " +
+            "Period = " + (_period == Timeout.UnsignedInfinite ? "(not set)" : TimeSpan.FromMilliseconds(_period).ToString("g")) + ", " +
+            _timerCallback.Method;
```
Comment (davidfowl) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:499]: What does this look like for Cancellation Tokens and Task.Delay?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs]: image

**Comment by r-ramesh** [comment]
@stephentoub Investigating this a bit more, is there a way to avoid using TickCount/TickCount64?
The reason is that this information is not available in a minidump. DateTime.Now is available (the time the dump was written), but not TickCount/TickCount64. As a result, expressions involving TickCount won't work on dumps.
Currently, in minidumps, the VS debugger just evaluates TickCount in the current debugger context, which is wrong and instead should spew an error that we can't evaluate Environment.TickCount. DateTime.Now is also currently evaluated in the current context but that's fixable to instead plumb through the time when the minidump was collected.

**Comment by stephentoub** [comment]
> Is there a way to avoid using TickCount/TickCount64?

No.  It's the only way to make sense of the values stored in the timer instance.

**Comment by r-ramesh** [comment]
> > Is there a way to avoid using TickCount/TickCount64?
> 
> No. It's the only way to make sense of the values stored in the timer instance.

I see. Do you have any suggestions on how to view this in minidumps? I imagine that is the scenario where a visualization might be most helpful.
For example, the timers extension just displays the DueTime as is [https://github.com/dotnet/diagnostics/blob/6f0320dc62fb22dca96b652a94fced839e0e660f/src/Microsoft.Diagnostics.ExtensionCommands/TimersCommand.cs#L87](https://github.com/dotnet/diagnostics/blob/6f0320dc62fb22dca96b652a94fced839e0e660f/src/Microsoft.Diagnostics.ExtensionCommands/TimersCommand.cs#L87)
Is that number useful as is?

**Comment by stephentoub** [comment]
> Is that number useful as is?

That number tells you the configured timeout for the timer, e.g. if I do `new Timer(_ => { }, null, 12345, 6789)`, the number will be approximately 12345 if this is the first time the timer was scheduled, or it'll be approximately 6789 thereafter.  But it doesn't tell you how long until the timer fires.  The above timer could have been created 12344 milliseconds ago, such that it's going to fire in 1ms, but it'll still show you 12345.

> Do you have any suggestions on how to view this in minidumps?

Does a DebuggerDisplay / DebuggerTypeProxy have the ability to know whether it's live or a dump?  What happens if I just try/catch around the attempt to use TickCount?

**Comment by stephentoub** [comment]
There doesn't appear to be an easy way to detect in the DebuggerTypeProxy/DisplayString that we're dump debugging, nor does it appear possible to eat an exception that's thrown and handle it gracefully.  Which is unfortunate, given that the code is unable to use a bunch of things it would otherwise want to, like Monitor.TryEnter (Monitor.Enter hangs the debugger if another thread is holding the lock), TimeSpan.ToString("g"), TickCount, etc.  I pushed a change that surfaces the next firing time only in the type proxy so that the other information is still valid, that enumerates the timers in a thread-unsafe way, etc.  Far from ideal, but maybe a reasonable compromise given the debugger limitations.

**Comment by stephentoub** [comment]
One more change.  I worked around the inability to query the current time by storing a one-time mapping between tick count and UTC time when the timer system is initialized.  That can then be used to interpret the tick counts in timers. It's not perfect, but it's something.

**Comment by stephentoub** [comment]
@kouvel, any concerns?

**Review by davidfowl — APPROVED** [review]
LGTM, clever trick with the estimated next utc time.

**Review by kouvel — APPROVED** [review]
LGTM

**Comment by r-ramesh** [comment]
@stephentoub 👍

**Comment by stephentoub** [comment]
Thanks for the reviews.

**Pull request merged**